### ISSUE TITLE
[BUG] Improve `cpp2::in` (restore old behavior, exclude problem cases only)

### ISSUE THREAD

**Issue by JohelEGP** [hsutter/cppfront#317]
**Describe the bug**

> https://github.com/hsutter/cppfront/issues/248#issuecomment-1493069838 finally pushed to comment on this. This formulation:
> - Passes fundamental types of size greater than 2 object pointers, like some 256-bit extended integer type.
> - Introduces a regression. Types that are not fundamental due to ABI issues, like GCC's 128-bit integer types (`-pedantic`), are passed by `const&`.
>
> I suggest reverting the deletion of `prefer_pass_by_value` and enhancing it to exclude the problematic [incomplete types](https://en.cppreference.com/w/cpp/language/type#Incomplete_type) that can affect `cpp2::in`.
>
> -- https://github.com/hsutter/cppfront/commit/a89af8fcf1c258fb03ea17947eb76b1f6e410a25#r107115865

Also fix `<` to `<=` to consider types exactly the size of 2 object pointers. See https://compiler-explorer.com/z/8EKPMrhee.

**Comment by JohelEGP** [comment]
Here's a visualization of the `in` types taken by value vs. by reference
before a89af8f, since a89af8f, and proposed here (#317),
assuming
- the "fix `<` to `<=`" above, and
- an implementation with
  * `sizeof(void*) == 8`, 64 bits, and
  * some hypothetical integer types.

```Cpp2
i8 i16 i32 i64
u8 u16 u32 u64
i128 i256 // Extended integer types.
u128 u256
__i128 __i256 // Non-fundamental integer types.
__u128 __u256
enum enum16  : std::int16_t  { }; // C++1
enum enum256 : std::int256_t { }; // C++1
my_t: type = { }
```

| a89af8f] val | a89af8f] `&` | [a89af8f val | [a89af8f `&` | #317 val | #317 `&`  |
|:-------------|:-------------|:-------------|:-------------|:---------|:----------|
| `i8`         |              | `i8`         |              | `i8`     |           |
| `i16`        |              | `i16`        |              | `i16`    |           |
| `i32`        |              | `i32`        |              | `i32`    |           |
| `i64`        |              | `i64`        |              | `i64`    |           |
| `i128`       |              | `i128`       |              | `i128`   |           |
|              | `i256`       | `i256` ❌     |              |          | `i256`    |
| `u8`         |              | `u8`         |              | `u8`     |           |
| `u16`        |              | `u16`        |              | `u16`    |           |
| `u32`        |              | `u32`        |              | `u32`    |           |
| `u64`        |              | `u64`        |              | `u64`    |           |
| `u128`       |              | `u128`       |              | `u128`   |           |
|              | `u256`       | `u256` ❌     |              |          | `u256`    |
| `__i128`     |              |              | `__i128` ❌   | `__i128` |           |
|              | `__i256`     |              | `__i256`     |          | `__i256`  |
| `__u128`     |              |              | `__u128` ❌   | `__u128` |           |
|              | `__u256`     |              | `__u256`     |          | `__u256`  |
| `enum16`     |              | `enum16`     |              | `enum16` |           |
|              | `enum256`    | `enum256` ❌  |              |          | `enum256` |
| `my_t` ❌     |              |              | `my_t` ✔     |          | `my_t` ✔  |

a89af8f's commit message mentions:
> Changed in\<T> to pass fundamental types by value. Only those are known to be always complete. Partly addresses #270.

In no discussion was it even hinted at to also disregard the size part from the convention:
> * (Simple) ((Foundation)) Warn when a parameter being passed by value has a size greater than `2 * sizeof(void*)`.
  Suggest using a reference to `const` instead.
> -- https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rf-in

As shown in the table, since a89af8f,
- size is not taken into account, so types of size greater than `2 * sizeof(void*)` are passed by value, and
- types that can never be incomplete are always passed by `const&` even if their size is not greater than `2 * sizeof(void*)`.

**Comment by JohelEGP** [comment]
Based on the discussion at https://github.com/hsutter/cppfront/commit/a89af8fcf1c258fb03ea17947eb76b1f6e410a25#r107115865, I propose this new formulation for `cpp2::in`:
```C++
template<typename T>
constexpr bool prefer_pass_by_value =
    sizeof(T) <= 2*sizeof(void*)
    && std::is_trivially_copy_constructible_v<T>;

template<typename T>
    requires (std::is_enum_v<T> && !std::is_scoped_enum_v<T>)
             || std::is_class_v<T>
             || std::is_union_v<T>
constexpr bool prefer_pass_by_value<T> = false;

template<typename T>
using in =
    std::conditional_t<
        prefer_pass_by_value<T>,
        T const,
        T const&
    >;
```

**Comment by gregmarr** [comment]
Unfortunately, it doesn't look like `is_trivially_copy_constructible_v` can be used here:

https://en.cppreference.com/w/cpp/types/is_copy_constructible
> T shall be a complete type, (possibly cv-qualified) void, or an array of unknown bound. Otherwise, the behavior is undefined.

**Comment by JohelEGP** [comment]
You must exclude the cases where `T` can be incomplete without `sizeof` and `std::is_trivially_copy_constructible_v` requiring to be valid, as those require a complete type. See https://compiler-explorer.com/z/8ejEG9qqK.

**Comment by gregmarr** [comment]
So that's the reason for the two versions, there's no short-circuiting there, and `is_class_v` and `is_union_v` covers everything that can be an incomplete type?

**Comment by JohelEGP** [comment]
See https://github.com/hsutter/cppfront/commit/a89af8fcf1c258fb03ea17947eb76b1f6e410a25#r107716660. Can you prove they affect `cpp2::in`?

**Comment by filipsajdak** [comment]
@JohelEGP I have done a small test with your proposed solution: https://godbolt.org/z/qnYaYcxPM

```cpp
auto fun(in<int>) {} // works
auto fun(in<S>)   {} // works
auto fun(in<C>)   {} // works
auto fun(in<E>)   {} // works
auto fun(in<U>)   {} // works

auto fun(in<A>)   {} // fails, but works with std::is_scalar_v
auto fun(in<AS>)  {} // fails, but works with std::is_scalar_v
auto fun(in<V>)   {} // fails, need to add requires (!std::is_void_v<T>)
```

**Comment by filipsajdak** [comment]
We can only try to declare them: https://godbolt.org/z/sn11YdY6f
```cpp
auto fun(in<int>); // works
auto fun(in<S>);   // works
auto fun(in<C>);   // works
auto fun(in<E>);   // works
auto fun(in<U>);   // works

auto fun(in<A>);   // fails, but works with std::is_scalar_v
auto fun(in<AS>);  // fails, but works with std::is_scalar_v
auto fun(in<V>);   // fails, need to add requires (!std::is_void_v<T>)
```

The types might comes from cpp1 we might don't know what they are alias for, or it just might be inside some generic code (that is why I have tested it against `void`).

**Comment by filipsajdak** [comment]
TLDR: currently (probably) the most reliable version is:

```cpp
template<typename T>
    requires (!std::is_void_v<T>)
using in =
    std::conditional_t<
        std::is_scalar_v<T>,
        T const,
        T const&
    >;
```

**Comment by JohelEGP** [comment]
> The types might comes from cpp1 we might don't know what they are alias for, or it just might be inside some generic code (that is why I have tested it against `void`).

You _may_ be right about `void`. It's actually legal, but I'll leave it to Herb to decide how to handle that.

See https://compiler-explorer.com/z/PKdWo11qe.
```C++
auto fun(in<A>)   {} // works
auto fun0(A) { }
void g(A a) { fun0(a); } // works

auto fun(in<AS>)  {} // works? should it?
auto fun1(AS) { }
void g(AS a) { fun1(a); } // works

auto fun(in<V>)   {} // works, should it?
auto fun2(void) { } // works
void g() { fun2(); }
```

So those work, but not when going through `cpp2::in` as previously defined. I think we need to unconditionally include arrays of unknown bound, as mentioned at the end of https://github.com/hsutter/cppfront/commit/a89af8fcf1c258fb03ea17947eb76b1f6e410a25#r107715926. I'm not so sure about the semantics of bounded arrays as arguments, but they seem to work, so they could be conditionally included. I also included `void` because it's legal.

**Comment by JohelEGP** [comment]
I thought a bit more. `in void` doesn't make sense, so let's take your suggestion of preventing that with a _requires-clause_. I fixed `prefer_pass_by_value` to `true` for arrays. I think they decay to pointers as parameters. See the mangled signatures for `fun0` and `fun1`:
- GCC `fun0(int*)` `fun1(S*)`
- Clang `fun0(int*)` `fun1(S*)`
- MSVC `fun0(int * const)` `fun1(S * const)`

I'll amend https://github.com/hsutter/cppfront/issues/317#issuecomment-1498135632. See https://compiler-explorer.com/z/K1TsT7aeG.

**Comment by JohelEGP** [comment]
As a parameter,
- an array decays to a pointer, and
- a reference to an array remains a reference to an array.

`in x: T` declares the intent to take a `T` as a parameter. If `T` is an array, `x` should be (a reference to) an array. A pointer is not an array, and `cpp2::in<T>` should reflect this.

I'll amend https://github.com/hsutter/cppfront/issues/317#issuecomment-1498135632. See https://compiler-explorer.com/z/TEe98e171.

**Comment by JohelEGP** [comment]
Made https://github.com/hsutter/cppfront/issues/317#issuecomment-1496793260's table shorter and nicer, and added:
| Before a89af8f     | Since a89af8f      | #317             |
|:-------------------|:-------------------|:-----------------|
| `my_t[]` ❌ (error) | `my_t[]` `&` ✔     | `my_t[]` `&` ✔   |
| `my_t[3]` `&`      | `my_t[3]` `&`      | `my_t[3]` `&`    |

**Comment by JohelEGP** [comment]
https://en.cppreference.com/w/cpp/language/function#Parameter_list mentions
![1680781658](https://user-images.githubusercontent.com/21071787/230367664-1ac99460-d977-44c0-a4b3-857cfe2a2468.png)

As expected, a function type broke `sizeof`: https://compiler-explorer.com/z/chTdn4zMb.

Following the logic of https://github.com/hsutter/cppfront/issues/317#issuecomment-1498340096, `in x: T` where `T` is `void()`, should take the function by reference. A function parameter also decays to a pointer. It doesn't lose size information like an array. But a pointer is nullable and an object, so it should be opt-in.

I'll amend https://github.com/hsutter/cppfront/issues/317#issuecomment-1498135632 and https://github.com/hsutter/cppfront/issues/317#issuecomment-1496793260. See https://compiler-explorer.com/z/GevMPo65K.

| Before a89af8f     | Since a89af8f      | #317             |
|:-------------------|:-------------------|:-----------------|
| `void()` ❌ (error) | `void()` `&` ✔     | `void()` `&` ✔   |

**Comment by JohelEGP** [comment]
Thank you everyone for participating. I think my solution is ready, and much better that what I could've cooked up alone. I'll leave it to Herb to decide what to take from my and other's raised points.

**Comment by filipsajdak** [comment]
Is there a PR somewhere? I can test it “in the field”.

**Comment by JohelEGP** [comment]
You can apply this git-patch.
```diff
diff --git a/include/cpp2util.h b/include/cpp2util.h
index 23a78bc..16dd1f4 100644
--- a/include/cpp2util.h
+++ b/include/cpp2util.h
@@ -492,9 +492,19 @@ template<typename T>
 //-----------------------------------------------------------------------
 //
 template<typename T>
+constexpr bool prefer_pass_by_value =
+    sizeof(T) <= 2*sizeof(void*)
+    && std::is_trivially_copy_constructible_v<T>;
+
+template<typename T>
+    requires std::is_class_v<T> || std::is_union_v<T> || std::is_array_v<T> || std::is_function_v<T>
+constexpr bool prefer_pass_by_value<T> = false;
+
+template<typename T>
+    requires (!std::is_void_v<T>)
 using in =
-    std::conditional_t <
-        std::is_scalar_v<T>,
+    std::conditional_t<
+        prefer_pass_by_value<T>,
         T const,
         T const&
     >;
```

**Comment by hsutter** [comment]
Thanks! I've taken a slight variation of this -- please check the commit, and please reopen this issue if I missed something.

(General note: I'm distracted with the upcoming WG21 Varna meeting and will be slow to follow up on issues/PRs until after that.)

**Comment by JohelEGP** [comment]
The separate specialization was to avoid using `sizeof` when `in` is passed an incomplete type.
Will that not be a problem?

**Comment by hsutter** [comment]
But we still know whether the incomplete type is a class, union, or function type, correct? That information has to be part of the forward declaration.

**Comment by hsutter** [comment]
Correction: Ah, now I see the problem. Thanks! Pushing a fix to go back to your way of doing it...
